html/captive-portal/lib/captiveportal/PacketFence/Controller/Billing.pm

```diff
@@ -0,0 +1,383 @@
+    my ($self, $c) = @_;
+    $c->forward('validate');
+    my $billing = $c->stash->{billing};
+    my $data;
+    eval {
+        $data =
+          $billing->prepare_payment($c->session, $c->stash->{tier}, $c->request->parameters, $c->request->path);
+    };
+    if ($@) {
+        $c->log->error($@);
+        $c->stash(template => 'billing/index.html');
+        $c->detach('index');
+    }
+    $c->stash($data);
+    $c->stash(template => "billing/confirm_" . $billing->type . ".html");
+}
+
+=head2 validate
+
+=cut
+
+sub validate : Private {
+    my ($self, $c) = @_;
+    my $request = $c->request;
+    my $source_param = first { /^billing_source_/ } $request->param_names;
+    if($source_param =~ /^billing_source_(.*)/) {
+        $self->source($c, $1);
+    } else {
+        $self->showError($c, "Invalid billing source for profile");
+    }
+    #Check if the billing source provided is correct
+    my $selected_tier = $request->param('tier');
+    my $first_name = $request->param('firstname');
+    my $last_name = $request->param('lastname');
+    my $email;
+    my $person;
+    if($c->session->{username}) {
+        $person = person_view($c->session->{username});
+        $email = $c->session->{username}
+    }
+    else {
+        $email = $request->param('email');
+    }
+    unless ($selected_tier) {
+        $c->log->error("No Tier selected");
+        $c->stash({
+            template => 'billing/index.html',
+            txt_validation_error => 'No Tier selected',
+        });
+        $c->detach('index');
+    }
+    my $tier = $c->profile->findTier($selected_tier);
+    unless ($tier) {
+        $c->log->error("Selected Tier is invalid");
+        $c->stash({
+            template => 'billing/index.html',
+            txt_validation_error => 'Selected Tier is invalid',
+        });
+        $c->detach('index');
+    }
+    my $valid_name = ( pf::web::util::is_name_valid($first_name)
+            && pf::web::util::is_name_valid($last_name) );
+    my $valid_email = pf::web::util::is_email_valid($email);
+    my %temp = (
+        "firstname" => $first_name,
+        "lastname"  => $last_name,
+        "email"     => $email,
+        "username"  => $email,
+        "tier"      => $tier,
+    );
+    $c->session(%temp);
+    $c->stash(%temp);
+}
+
+=head2 cancel
+
+=cut
+
+sub cancel : Chained('source') : Args(0) {
+    my ($self, $c) = @_;
+    my $request = $c->request;
+    my $billing = $c->stash->{billing};
+    my $data;
+    eval {
+        $data = $billing->cancel($c->session, $request->parameters, $request->path);
+    };
+    if ($@) {
+        $c->log->error($@);
+    }
+    else {
+        $c->stash($data);
+    }
+    $c->stash({
+        template => 'billing/index.html',
+        txt_validation_error => 'Order was canceled',
+    });
+    $c->detach('index');
+}
+
+=head2 index
+
+=cut
+
+sub index : Path : Args(0) {
+    my ($self, $c) = @_;
+    my @billing_sources = $c->profile->getBillingSources;
```

Comment:
This line and the 3 below should have the logic encapsulated in the profile or at least in another method for reusal